File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[876]链表的中间结点.scala
package leetcode.editor.cn
object MiddleOfTheLinkedList {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def middleNode(head: ListNode): ListNode = {
      var fast = head
      var slow = head
      while(fast != null && fast.next != null){
        slow = slow.next
        fast = fast.next.next
      }
      slow
    }
}
//leetcode submit region end(Prohibit modification and deletion)

}

File: cpp/leetcodecn/[876]链表的中间结点.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <iostream>


using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *middleNode(ListNode *head) {
        auto fast = head;
        auto slow = head;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
    Solution solution;
    auto result = solution.middleNode(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5,new ListNode(6)))))));
    cout << result->val << endl;
}